File: api/client/src/main/java/su/plo/voice/api/client/audio/device/AlAudioDevice.java
package su.plo.voice.api.client.audio.device;

import org.jetbrains.annotations.NotNull;

import java.util.Optional;

// todo: doc
public interface AlAudioDevice extends AudioDevice {

    /**
     * Gets the device's pointer
     */
    Optional<Long> getPointer();

    /**
     * Gets the device's context pointer
     */
    Optional<Long> getContextPointer();

    /**
     * Runs runnable in the device's context
     */
    default void runInContext(@NotNull DeviceRunnable runnable) {
        runInContext(runnable, true);
    }

    void runInContext(@NotNull DeviceRunnable runnable, boolean blocking);

    interface DeviceRunnable {

        void run() throws DeviceException;
    }
}

File: api/client/src/main/kotlin/su/plo/voice/api/client/audio/device/AlAudioDevice.kt
package su.plo.voice.api.client.audio.device

import kotlinx.coroutines.Runnable
import java.util.concurrent.CompletableFuture

// todo: doc
interface AlAudioDevice : AudioDevice {

    /**
     * Gets the device's pointer
     */
    val devicePointer: Long

    /**
     * Gets the device's context pointer
     */
    val contextPointer: Long

    /**
     * Runs runnable in the device's context
     */
    @JvmSynthetic
    suspend fun runInContext(runnable: suspend () -> Unit)

    /**
     * Runs runnable in the device's context
     */
    fun runInContextAsync(runnable: Runnable): CompletableFuture<Void?>

    /**
     * Runs runnable in the device's context
     */
    fun runInContextBlocking(runnable: Runnable) = runInContextAsync(runnable).get()
}
